### DISCUSSION TITLE
RFC: Have OpenTTD *not* lock the cursor position when dragging viewports by default

### DISCUSSION THREAD

**Discussion by embeddedt** [OpenTTD/OpenTTD#9150] [Features]
OpenTTD has historically locked the mouse cursor position when the viewport is being dragged. This feature works on many systems without problems. However, on some platforms (e.g. Emscripten), pointer locking is difficult to implement without drawbacks. In particular, interacting with the Emscripten port currently requires the pointer to be locked inside the browser window all the time, which prevents moving to other applications easily and swiftly.

I explored a workaround for Emscripten in #8654 but the solution there had its own problems.

In that PR, @TrueBrain noted that there is actually an option to allow dragging the viewport without locking the mouse position. It seems that this option was added in #6756 to solve precisely the kind of issue that the Emscripten port is currently having.

The simplest solution would therefore be to make the non-locking option the default - this would make the Emscripten port work intuitively out-of-box, while still retaining legacy behavior for those who find it important.

Some alternatives:

* Remove support for locking the cursor altogether.
* Make this mode default, but only on the Emscripten port.

**Comment by embeddedt** [comment]
#9191 has been merged.

**Comment by LC-Zorg** [comment]
Due to niche applications, do you want the game to always use such a cumbersome solution by default? :(

It's a matter of preference, but I don't really know a game that uses this method of changing views. They may be, but in OTTD it just won't work. It would be tedious especially with larger maps. To get from point A to point B the player will have to repeatedly move the mouse back and forth using 1/4 of the possible wrist / mouse movement - sorry, but that won't be any improvement over what is now.

> Some alternatives:

> 1. Remove support for locking the cursor altogether.

Terrifying

> 2. Make this mode default, but only on the Emscripten port.

Very sensible

3. Addition of a vector method of changing the view

I admit that at the moment I don't remember a game or a program that would use it, although it can be quite popular (Word and Excel allow you to scroll the view this way by pressing the scroll wheel). I also don't know if the name is correct. ;)
This is a solution that, above all, would significantly facilitate moving around large maps, and also allow you to move smoothly around the map. I'm not sure about that, but it might also be a solution to the problem with emscripten.

**What is it about and how would it work?**
1. Press RMB on the map (the game saves the cursor coordinates on the screen)
2. You move the mouse - the cursor changes its position (but it is not related to the terrain point)
3. The difference between the written coordinates of the cursor position and the current ones determines the displacement vector
4. The longer the vector, the greater the speed of displacement
5. This method allows both very slow and very fast movements

**Do you want to test this solution?**
You can now pan the view in a very similar way by pressing and holding the C key:
+ position the cursor in the center of the screen
+ press and hold C
+ move cursor
The view moves smoothly at any speed - you can fly the balloon over mountains and cities, or you can find yourself on the other side of the 4k map in 1 second without having to open the minimap or repeatedly dragging the mouse left and right on the table.

Perhaps it is possible to code a function where pressing RMB would work similarly (not the same) as pressing C, but without the nervousness of the movement?

**Reply by embeddedt** [reply]
Thanks for the feedback. :slightly_smiling_face: 

> To get from point A to point B the player will have to repeatedly move the mouse back and forth using 1/4 of the possible wrist / 
mouse movement - sorry, but that won't be any improvement over what is now.

Hmm... with my setup it makes very little difference, as I don't have enough space to move my mouse further than the distance it takes to go across the screen once anyways. So for me I still have to pick up the mouse and place it back down whether the pointer is locked or not.

Additionally, how many people scroll across the map in this manner? Personally I almost never use the mouse to move around the map. I prefer to use arrow keys, or better yet, look up a station in the general area and jump to it.

> Perhaps it is possible to code a function where pressing RMB would work similarly (not the same) as pressing C, but without the nervousness of the movement?

So, in essence, you are suggesting that we use the solution in #9191, but with the addition of inertial scrolling so that it is still possible to efficiently move across large areas? That sounds doable (pending core devs' feedback, of course).

**Comment by LC-Zorg** [comment]
> Hmm... with my setup it makes very little difference, as I don't have enough space to move my mouse further than the distance it takes to go across the screen once anyways. So for me I still have to pick up the mouse and place it back down whether the pointer is locked or not.

And the point is that with the solution you propose, by moving the cursor across the entire diagonal of the screen, you only make a small movement with your hand - you don't use all the hand movement you have and to move a greater distance on map you have to repeatedly drag the mouse left and right. With the current default setting, you can use full hand motion and save that much waving.

> Personally I almost never use the mouse to move around the map. I prefer to use arrow keys, or better yet, look up a station in the general area and jump to it.

8O Why use arrows when you are holding the mouse anyway* and your fingers are on the buttons? :D
*Unless you are not using a mouse at all or are using a touchpad.

> how many people scroll across the map in this manner?

I don't know the statistics, but I suppose that the vast majority of them play with the mouse, and then using arrows is only unnecessarily engaging and limiting.
Besides, the arrows only give you 8 directions of movement (the mouse gives you unlimited amount) and moving them is terribly slow (again, the mouse allows you to change the view at any speed).

> So, in essence, you are suggesting that we use the solution in #9191, but with the addition of inertial scrolling so that it is still possible to efficiently move across large areas? That sounds doable (pending core devs' feedback, of course).

I don't know if it's the same for sure. :) Rather, I was thinking about an additional option that could actually possibly be the default. I sure wouldn't want to delete the current default method.

Little visualization ;)
![Displacement speed vector v1 1](https://user-images.githubusercontent.com/67662046/120896187-bf8eec00-c620-11eb-8546-6804d16e5e0b.png)
- holding RMB (optionally LMB) you can constantly change the direction and speed of movement.
- releasing RMB you stop the view change
- by returning the cursor to the point / zone "0" - you also stop the change of view

**Reply by embeddedt** [reply]
When I started playing OpenTTD I only had a touchpad, so the arrows were much easier to use. That's probably why I'm so used to them. :laughing:
